in-market-test/v2/ansible/parse_logs.py

```diff
@@ -0,0 +1,194 @@
+import re
+from collections import Counter
+from dataclasses import dataclass
+from pathlib import Path
+
+time_multiplier = {
+    "s": 1_000_000,
+    "ms": 1_000,
+    "µs": 1,
+}
+
+
+def time_string(µs):
+    # microsecond * 1,000,000 = 1 second
+    if µs < 1_000:
+        return f"{µs}µs"
+    if µs < 1_000_000:
+        return f"{µs/1_000}ms"
+    s = µs / 1_000_000
+    if s < 60:
+        return f"{s:.2f}s"
+    if s < 3600:
+        return f"{int(s//60)}m{s % 60:.1f}s"
+    return f"{int(s//3600)}h{int(s//60) % 60}m{s % 60:.1f}s"
+
+
+def split_logs_into_queries(log_file: Path):
+    if not log_file.exists():
+        return []
+    full_log_contents = log_file.read_text()
+    queries_log_contents = full_log_contents.split(
+        "ipa_core::query::runner::oprf_ipa: new"
+    )
+    if len(queries_log_contents) < 2:
+        return []
+    return queries_log_contents[1:]
+
+
+@dataclass
+class Step:
+    size: int
+    method: str
+    busy_time: float
+    idle_time: float
+
+    @property
+    def total_time(self):
+        return self.busy_time + self.idle_time
+
+    def percent_busy_time(self, total_busy_time):
+        return 100.0 * self.busy_time / total_busy_time
+
+    def percent_idle_time(self, total_idle_time):
+        return 100.0 * self.idle_time / total_idle_time
+
+    def percent_total_time(self, total_time):
+        return 100.0 * self.total_time / total_time
+
+
+@dataclass
+class Steps:
+    steps: list[Step]
+    helper: int
+
+    @classmethod
+    def build_from_logs(cls, helper: int, log_contents: str):
+        pattern = (
+            r".*{sz=(\d+)}:([a-zA-Z_]+):.*close time\.busy=(\d+\.?\d*)(ms|µs|s) "
+            r"time\.idle=(\d+\.?\d*)(ms|µs|s)"
+        )
+        matches = re.findall(pattern, log_contents)
+        steps = [
+            Step(
+                size=int(re_match[0]),
+                method=re_match[1],
+                busy_time=float(re_match[2]) * time_multiplier[re_match[3]],
+                idle_time=float(re_match[4]) * time_multiplier[re_match[5]],
+            )
+            for re_match in matches
+        ]
+        return cls(
+            steps=steps,
+            helper=helper,
+        )
+
+    @property
+    def total_busy_time(self):
+        return sum(step.busy_time for step in self.steps)
+
+    @property
+    def total_idle_time(self):
+        return sum(step.idle_time for step in self.steps)
+
+    @property
+    def total_time(self):
+        return sum(step.total_time for step in self.steps)
+
+    @property
+    def size(self):
+        if not self.steps:
+            return None
+        return self.steps[0].size
+
+    @property
+    def report(self):
+        if len(self.steps) == 0:
```

Comment:
can we use a library to create the table? or at least separate the "table printing" logic from Steps?

Reply:
I'm hoping to avoid an libraries - this should be able to run with just python3. 

as for separating it, this is just a quick and dirty script to use when testing. not trying to get too fancy.

Reply:
Can we move this to a separate module function?  (same with errors): `steps_to_report(steps: Steps)` and `errors_to_report(...)`

There's so much going on in these methods I don't think they should be marked as @property.